### ISSUE TITLE
Bazel query hangs when using allrdeps(siblings(rbuildfiles(...))) query in certain situations

### ISSUE THREAD

**Issue by luna-duclos** [bazelbuild/bazel#13359]
### Description of the problem / feature request:

Bazel query hangs when running certain queries, they look as follows (snipped as this only happens in a large monorepo):

```bash
bazel query \
  --universe_scope="//..." \
  --order_output=no \
  --keep_going \
  "(siblings(rbuildfiles($build_files_with_commas)))
  union (allrdeps(set(${changed_files[@]})))
  union (allrdeps(siblings(rbuildfiles($build_files_with_commas))))"
```

The query runs just fine when removing the allrdeps(siblings(rbuildfiles(...))) part of it

### Bugs: what's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.

I unfortunately haven't been able to get a reliable repro for this, I did however take a stack dump of the hung bazel server and attached it to this issue.

### What operating system are you running Bazel on?

Linux, or more specifically, a docker container running in k8s

### What's the output of `bazel info release`?

```
INFO: Invocation ID: 958fa642-af62-4518-aa64-29aca6b626e7
release 4.0.0
```

### What's the output of `git remote get-url origin ; git rev-parse master ; git rev-parse HEAD` ?

The repo this happens in is sadly not public, so not relevant

###  Have you found anything relevant by searching the web?

I have sadly not

### Any other information, logs, or outputs that you want to share?

[stackdump.txt](https://github.com/bazelbuild/bazel/files/6316370/stackdump.txt)

**Comment by meisterT** [comment]
Can you check whether the JSON trace profile reveals some useful insight? https://docs.bazel.build/versions/master/skylark/performance.html#performance-profiling

**Comment by luna-duclos** [comment]
Will the profiling work if I kill the process ? Since it hangs forever, happy to try it if so @meisterT

**Comment by meisterT** [comment]
Yes, it should. Just press Ctrl+C after a while and then load the profile. It should write what's available up to that point.

**Comment by michaeledgar** [comment]
The issue comes from a long-standing deficiency in the query engine's threading model. I've filed issue #13682 to track the broader issue with SkyQuery hangs.

Based on the stack trace you provided (thank you for including it!) you're suffering from a priority inversion: a shared thread pool is using all of its threads on the `siblings` function when the `rbuildfiles` function has not completed.

This rewritten query is equivalent, and should force `rbuildfiles` to finish evaluation before `siblings` evaluation begins, and force `siblings` to complete before `allrdeps` begins:

```
let rb = rbuildfiles($build_files_with_commas) in
  let sib = siblings($rb) in
    $sib + allrdeps($sib + set(${changed_files[@]}))
```

**Comment by BenHenning** [comment]
Thanks @michaeledgar that's really helpful! I'll also follow up on the new issue.
